using_restricted_array.rb

```diff
@@ -38,29 +72,78 @@ def sort(array, length)
 # constant, adds an element with 'SPECIAL_VALUE' in the end. Assumes the array
 # to be sorted in ascending order.
 def delete(array, length, value_to_delete)
-  puts "NOT IMPLEMENTED"
+  i = 0
+
+  until i == length do
+    if array[i] == value_to_delete
+      array[i] = SPECIAL_VALUE
+    end
+    i += 1
+  end
 end
 
 # Restricted array cannot be resized. So, we workaround by having a convention
 # Convention: replace all values with 'SPECIAL_VALUE'
 # Empties the restricted array by making all values = SPECIAL_VALUE
 def empty(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+
+  until i == length do
+    array[i] = SPECIAL_VALUE
+    i += 1
+  end
 end
 
 # Finds and returns the largest value element in the array which is not 'SPECIAL_VALUE'
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  max = array[0]
+  i = 0
+
+  until i == length do
+    if array[i] > max && array[i] != SPECIAL_VALUE
+      max = array[i]
+    end
+    i += 1
+  end
+
+  return max
 end
 
 # Insert value to insert at the correct index into the array assuming the array
 # is sorted in ascending manner.
 # Restricted arrays cannot be resized. Insert only if there is space in the array.
-# (Hint: if there are elements with 'SPECIAL_VALUE', there is no room to insert)
+# (Hint: if there are no elements with 'SPECIAL_VALUE', there is no room to insert)
 # All subsequent elements will need to be moved forward by one index.
 def insert_ascending(array, length, value_to_insert)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  index = nil
+  temp = nil
+
+  until i == length do
+    if array[i] >= value_to_insert
+      j = i
+      until j == length do
+        if array[j] == SPECIAL_VALUE
+          break
+        else
+          j += 1
+        end
+      end
+      if j == length
+        return "No Room to Insert"
+      else
+        until j == i
+          array[j] = array[j - 1]
+          j -= 1
+        end
+        array[i] = value_to_insert
+      end
+    end

```

Comment:
You need to increment i after the if statement. You currently have an infinite loop.